File: compat/suites/dotnet-sdk/Groups/SsmGroup.cs
using Amazon.SimpleSystemsManagement;
using Amazon.SimpleSystemsManagement.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class SsmGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["PutParameter"] = PutParameterAsync,
        ["GetParameter"] = GetParameterAsync,
        ["PutParameterOverwrite"] = PutParameterOverwriteAsync,
        ["GetParameterHistory"] = GetParameterHistoryAsync,
        ["PutMultipleParameters"] = PutMultipleParametersAsync,
        ["GetParameters"] = GetParametersAsync,
        ["DescribeParameters"] = DescribeParametersAsync,
        ["TagParameter"] = TagParameterAsync,
        ["ListSSMTagsForResource"] = ListSSMTagsForResourceAsync,
        ["DeleteParameters"] = DeleteParametersAsync,
        ["PutSecureStringParameter"] = PutSecureStringParameterAsync,
        ["GetSecureStringParameter"] = GetSecureStringParameterAsync,
        ["GetSecureStringWithoutDecryption"] = GetSecureStringWithoutDecryptionAsync,
        ["GetParametersByPath"] = GetParametersByPathAsync,
        ["GetParametersByPathRecursive"] = GetParametersByPathRecursiveAsync,
        ["GetParametersByPathPaginated"] = GetParametersByPathPaginatedAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["ssm-parameters"] = SetupParametersAsync,
        ["ssm-secure"] = SetupSecureAsync,
        ["ssm-path"] = SetupPathAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["ssm-parameters"] = TeardownParametersAsync,
        ["ssm-secure"] = TeardownSecureAsync,
        ["ssm-path"] = TeardownPathAsync,
    };

    // ---- helpers ----

    private async Task DeleteParamsIfExistsAsync(List<string> names)
    {
        try
        {
            await clients.SSM().DeleteParametersAsync(new DeleteParametersRequest { Names = names });
        }
        catch
        {
        }
    }

    private static string RequireSsmParam(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }

    // ---- setups ----

    private async Task SetupParametersAsync(TestContext context)
    {
        var name = $"{context.RunId}-ssm-param";
        await clients.SSM().PutParameterAsync(new PutParameterRequest
        {
            Name = name,
            Value = "v1",
            Type = ParameterType.String,
            Overwrite = false,
        });
        context.Set("SsmBaseParam", name);
    }

    private async Task SetupSecureAsync(TestContext context)
    {
        var name = $"{context.RunId}-ssm-secure";
        await clients.SSM().PutParameterAsync(new PutParameterRequest
        {
            Name = name,
            Value = "my-secret",
            Type = ParameterType.SecureString,
            Overwrite = false,
        });
        context.Set("SsmSecureParam", name);
    }

    private async Task SetupPathAsync(TestContext context)
    {
        var name1 = $"/{context.RunId}/ssm-path/a";
        var name2 = $"/{context.RunId}/ssm-path/b";
        var ssm = clients.SSM();
        await ssm.PutParameterAsync(new PutParameterRequest
        {
            Name = name1,
            Value = "value1",
            Type = ParameterType.String,
            Overwrite = false,
        });
        await ssm.PutParameterAsync(new PutParameterRequest
        {
            Name = name2,
            Value = "value2",
            Type = ParameterType.String,
            Overwrite = false,
        });
        context.Set("SsmPathParam1", name1);
        context.Set("SsmPathParam2", name2);
    }

    // ---- teardowns ----

    private async Task TeardownParametersAsync(TestContext context)
    {
        var names = new List<string>();
        var baseParam = context.GetString("SsmBaseParam");
        if (baseParam is not null) names.Add(baseParam);
        names.Add($"{context.RunId}-ssm-put");
        var new1 = context.GetString("SsmNewParam1");
        if (new1 is not null) names.Add(new1);
        var new2 = context.GetString("SsmNewParam2");
        if (new2 is not null) names.Add(new2);
        await DeleteParamsIfExistsAsync(names);
    }

    private async Task TeardownSecureAsync(TestContext context)
    {
        var names = new List<string>();
        var secureParam = context.GetString("SsmSecureParam");
        if (secureParam is not null) names.Add(secureParam);
        names.Add($"{context.RunId}-ssm-secure-put");
        await DeleteParamsIfExistsAsync(names);
    }

    private async Task TeardownPathAsync(TestContext context)
    {
        var names = new List<string>();
        var param1 = context.GetString("SsmPathParam1");
        if (param1 is not null) names.Add(param1);
        var param2 = context.GetString("SsmPathParam2");
        if (param2 is not null) names.Add(param2);
        await DeleteParamsIfExistsAsync(names);
    }

    // ---- ssm-parameters ----

    private async Task PutParameterAsync(TestContext context)
    {
        var name = $"{context.RunId}-ssm-put";
        await clients.SSM().PutParameterAsync(new PutParameterRequest
        {
            Name = name,
            Value = "put-value",
            Type = ParameterType.String,
            Overwrite = false,
        });
        var getResponse = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
        });
        Assertions.NotNull(getResponse.Parameter, "PutParameter: Parameter");
        Assertions.Equal("put-value", getResponse.Parameter.Value,
            $"PutParameter: expected put-value but was {getResponse.Parameter.Value} (runId={context.RunId})");
    }

    private async Task GetParameterAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmBaseParam");
        var response = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
        });
        Assertions.NotNull(response.Parameter, "GetParameter: Parameter");
        Assertions.Equal("v1", response.Parameter.Value,
            $"GetParameter: expected v1 but was {response.Parameter.Value} (runId={context.RunId})");
    }

    private async Task PutParameterOverwriteAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmBaseParam");
        await clients.SSM().PutParameterAsync(new PutParameterRequest
        {
            Name = name,
            Value = "v2",
            Type = ParameterType.String,
            Overwrite = true,
        });
        var getResponse = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
        });
        Assertions.NotNull(getResponse.Parameter, "PutParameterOverwrite: Parameter");
        Assertions.Equal("v2", getResponse.Parameter.Value,
            $"PutParameterOverwrite: expected v2 but was {getResponse.Parameter.Value} (runId={context.RunId})");
    }

    private async Task GetParameterHistoryAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmBaseParam");
        var response = await clients.SSM().GetParameterHistoryAsync(new GetParameterHistoryRequest
        {
            Name = name,
        });
        Assertions.NotNull(response.Parameters, "GetParameterHistory: Parameters");
        Assertions.GreaterThanOrEqual(2, response.Parameters.Count,
            $"GetParameterHistory: expected >= 2 versions but was {response.Parameters.Count} (runId={context.RunId})");
    }

    private async Task PutMultipleParametersAsync(TestContext context)
    {
        var name1 = $"{context.RunId}-ssm-new1";
        var name2 = $"{context.RunId}-ssm-new2";
        var ssm = clients.SSM();
        await ssm.PutParameterAsync(new PutParameterRequest
        {
            Name = name1,
            Value = "new1",
            Type = ParameterType.String,
            Overwrite = false,
        });
        await ssm.PutParameterAsync(new PutParameterRequest
        {
            Name = name2,
            Value = "new2",
            Type = ParameterType.String,
            Overwrite = false,
        });
        context.Set("SsmNewParam1", name1);
        context.Set("SsmNewParam2", name2);

        var getResponse = await ssm.GetParametersAsync(new GetParametersRequest
        {
            Names = [name1, name2],
        });
        Assertions.GreaterThanOrEqual(2, getResponse.Parameters.Count,
            $"PutMultipleParameters: expected >= 2 params but was {getResponse.Parameters.Count} (runId={context.RunId})");
    }

    private async Task GetParametersAsync(TestContext context)
    {
        var name1 = RequireSsmParam(context, "SsmNewParam1");
        var name2 = RequireSsmParam(context, "SsmNewParam2");
        var response = await clients.SSM().GetParametersAsync(new GetParametersRequest
        {
            Names = [name1, name2],
        });
        Assertions.GreaterThanOrEqual(2, response.Parameters.Count,
            $"GetParameters: expected >= 2 params but was {response.Parameters.Count} (runId={context.RunId})");
    }

    private async Task DescribeParametersAsync(TestContext context)
    {
        var response = await clients.SSM().DescribeParametersAsync(new DescribeParametersRequest
        {
            ParameterFilters =
            [
                new ParameterStringFilter
                {
                    Key = "Name",
                    Option = "Contains",
                    Values = [context.RunId],
                },
            ],
        });
        Assertions.GreaterThanOrEqual(1, response.Parameters?.Count ?? 0,
            $"DescribeParameters: expected >= 1 param but was {response.Parameters?.Count ?? 0} (runId={context.RunId})");
    }

    private async Task TagParameterAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmBaseParam");
        await clients.SSM().AddTagsToResourceAsync(new AddTagsToResourceRequest
        {
            ResourceType = ResourceTypeForTagging.Parameter,
            ResourceId = name,
            Tags =
            [
                new Tag { Key = "project", Value = "overcast" },
            ],
        });

        var listResponse = await clients.SSM().ListTagsForResourceAsync(new ListTagsForResourceRequest
        {
            ResourceType = ResourceTypeForTagging.Parameter,
            ResourceId = name,
        });
        Assertions.True(
            (listResponse.TagList ?? []).Any(t => t.Key == "project" && t.Value == "overcast"),
            $"TagParameter: tag project=overcast not found (runId={context.RunId})");
    }

    private async Task ListSSMTagsForResourceAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmBaseParam");
        var response = await clients.SSM().ListTagsForResourceAsync(new ListTagsForResourceRequest
        {
            ResourceType = ResourceTypeForTagging.Parameter,
            ResourceId = name,
        });
        Assertions.True(
            (response.TagList ?? []).Any(t => t.Key == "project"),
            $"ListSSMTagsForResource: tag project not found (runId={context.RunId})");
    }

    private async Task DeleteParametersAsync(TestContext context)
    {
        var baseParam = RequireSsmParam(context, "SsmBaseParam");
        var new1 = RequireSsmParam(context, "SsmNewParam1");
        var new2 = RequireSsmParam(context, "SsmNewParam2");
        var putParam = $"{context.RunId}-ssm-put";
        var names = new List<string> { baseParam, new1, new2, putParam };

        var deleteResponse = await clients.SSM().DeleteParametersAsync(new DeleteParametersRequest
        {
            Names = names,
        });
        Assertions.GreaterThanOrEqual(3, deleteResponse.DeletedParameters.Count,
            $"DeleteParameters: expected >= 3 deleted but was {deleteResponse.DeletedParameters.Count} (runId={context.RunId})");

        var describeResponse = await clients.SSM().DescribeParametersAsync(new DescribeParametersRequest
        {
            ParameterFilters =
            [
                new ParameterStringFilter
                {
                    Key = "Name",
                    Option = "Contains",
                    Values = [context.RunId],
                },
            ],
        });
        Assertions.True(
            (describeResponse.Parameters ?? []).All(p => !names.Contains(p.Name)),
            $"DeleteParameters: params still present after deletion (runId={context.RunId})");
    }

    // ---- ssm-secure ----

    private async Task PutSecureStringParameterAsync(TestContext context)
    {
        var name = $"{context.RunId}-ssm-secure-put";
        var response = await clients.SSM().PutParameterAsync(new PutParameterRequest
        {
            Name = name,
            Value = "secure-value",
            Type = ParameterType.SecureString,
            Overwrite = false,
        });
        Assertions.GreaterThan(0, response.Version ?? 0,
            $"PutSecureStringParameter: expected Version > 0 but was {response.Version} (runId={context.RunId})");

        var getResponse = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
            WithDecryption = true,
        });
        Assertions.NotNull(getResponse.Parameter, "PutSecureStringParameter: Parameter");
        Assertions.Equal("secure-value", getResponse.Parameter.Value,
            $"PutSecureStringParameter: expected secure-value but was {getResponse.Parameter.Value} (runId={context.RunId})");
    }

    private async Task GetSecureStringParameterAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmSecureParam");
        var response = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
            WithDecryption = true,
        });
        Assertions.NotNull(response.Parameter, "GetSecureStringParameter: Parameter");
        Assertions.Equal(ParameterType.SecureString, response.Parameter.Type,
            $"GetSecureStringParameter: expected SecureString but was {response.Parameter.Type} (runId={context.RunId})");
        Assertions.Equal("my-secret", response.Parameter.Value,
            $"GetSecureStringParameter: expected my-secret but was {response.Parameter.Value} (runId={context.RunId})");
    }

    private async Task GetSecureStringWithoutDecryptionAsync(TestContext context)
    {
        var name = RequireSsmParam(context, "SsmSecureParam");
        var response = await clients.SSM().GetParameterAsync(new GetParameterRequest
        {
            Name = name,
            WithDecryption = false,
        });
        Assertions.NotNull(response.Parameter, "GetSecureStringWithoutDecryption: Parameter");
        Assertions.True(
            response.Parameter.Value != "my-secret",
            $"GetSecureStringWithoutDecryption: plaintext returned for encrypted param (runId={context.RunId})");
    }

    // ---- ssm-path ----

    private async Task GetParametersByPathAsync(TestContext context)
    {
        var path = $"/{context.RunId}/ssm-path";
        var response = await clients.SSM().GetParametersByPathAsync(new GetParametersByPathRequest
        {
            Path = path,
            Recursive = false,
        });
        Assertions.GreaterThanOrEqual(1, response.Parameters.Count,
            $"GetParametersByPath: expected >= 1 param but was {response.Parameters.Count} (runId={context.RunId})");
    }

    private async Task GetParametersByPathRecursiveAsync(TestContext context)
    {
        var path = $"/{context.RunId}";
        var response = await clients.SSM().GetParametersByPathAsync(new GetParametersByPathRequest
        {
            Path = path,
            Recursive = true,
        });
        Assertions.GreaterThanOrEqual(2, response.Parameters.Count,
            $"GetParametersByPathRecursive: expected >= 2 params but was {response.Parameters.Count} (runId={context.RunId})");
    }

    private async Task GetParametersByPathPaginatedAsync(TestContext context)
    {
        var path = $"/{context.RunId}/ssm-path";
        var page1 = await clients.SSM().GetParametersByPathAsync(new GetParametersByPathRequest
        {
            Path = path,
            Recursive = false,
            MaxResults = 1,
        });
        Assertions.Equal(1, page1.Parameters.Count,
            $"GetParametersByPathPaginated page1: expected 1 but was {page1.Parameters.Count} (runId={context.RunId})");
        Assertions.NotBlank(page1.NextToken,
            $"GetParametersByPathPaginated page1: missing NextToken (runId={context.RunId})");

        var page2 = await clients.SSM().GetParametersByPathAsync(new GetParametersByPathRequest
        {
            Path = path,
            Recursive = false,
            MaxResults = 1,
            NextToken = page1.NextToken,
        });
        Assertions.GreaterThanOrEqual(1, page2.Parameters.Count,
            $"GetParametersByPathPaginated page2: expected >= 1 param but was {page2.Parameters.Count} (runId={context.RunId})");
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/SsmGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.services.ssm.SsmClient;
import software.amazon.awssdk.services.ssm.model.*;

import java.util.List;
import java.util.Map;

/**
 * SSM Parameter Store compatibility test group.
 *
 * <p>Groups: ssm-parameters, ssm-secure, ssm-path.
 */
public final class SsmGroup implements ServiceGroup {

    private final AwsClients clients;

    public SsmGroup(AwsClients clients) {
        this.clients = clients;
    }

    private SsmClient ssm() { return clients.ssm(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("PutParameter",                    this::putParameter),
                Map.entry("GetParameter",                    this::getParameter),
                Map.entry("PutParameterOverwrite",           this::putParameterOverwrite),
                Map.entry("GetParameterHistory",             this::getParameterHistory),
                Map.entry("PutMultipleParameters",           this::putMultipleParameters),
                Map.entry("GetParameters",                   this::getParameters),
                Map.entry("DescribeParameters",              this::describeParameters),
                Map.entry("TagParameter",                    this::tagParameter),
                Map.entry("ListSSMTagsForResource",          this::listSsmTagsForResource),
                Map.entry("DeleteParameters",                this::deleteParameters),
                Map.entry("PutSecureStringParameter",        this::putSecureStringParameter),
                Map.entry("GetSecureStringParameter",        this::getSecureStringParameter),
                Map.entry("GetSecureStringWithoutDecryption",this::getSecureStringWithoutDecryption),
                Map.entry("GetParametersByPath",             this::getParametersByPath),
                Map.entry("GetParametersByPathRecursive",    this::getParametersByPathRecursive),
                Map.entry("GetParametersByPathPaginated",    this::getParametersByPathPaginated)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.ofEntries(
                Map.entry("ssm-parameters", this::setupParameters),
                Map.entry("ssm-secure",     this::setupSecure),
                Map.entry("ssm-path",       this::setupPath)
        );
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.ofEntries(
                Map.entry("ssm-parameters", this::teardownParameters),
                Map.entry("ssm-secure",     ctx -> deleteParamSilently(ctx.getString("ssmSecureParam"))),
                Map.entry("ssm-path",       this::teardownPath)
        );
    }

    // ── ssm-parameters ────────────────────────────────────────────────────────

    private void setupParameters(TestContext ctx) throws Exception {
        String name  = "/" + ctx.runId() + "/param1";
        String name2 = "/" + ctx.runId() + "/param2";
        ssm().putParameter(r -> r.name(name).value("value1").type(ParameterType.STRING));
        ssm().putParameter(r -> r.name(name2).value("value2").type(ParameterType.STRING));
        ctx.set("ssmParam1", name);
        ctx.set("ssmParam2", name2);
    }

    private void teardownParameters(TestContext ctx) {
        deleteParamSilently(ctx.getString("ssmParam1"));
        deleteParamSilently(ctx.getString("ssmParam2"));
    }

    private void putParameter(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("ssmParam1"), "ssmParam1");
    }

    private void getParameter(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmParam1");
        var resp = ssm().getParameter(r -> r.name(name));
        Assertions.assertEquals("value1", resp.parameter().value(), "GetParameter: value mismatch");
    }

    private void putParameterOverwrite(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmParam1");
        ssm().putParameter(r -> r.name(name).value("updated1").type(ParameterType.STRING).overwrite(true));
        var resp = ssm().getParameter(r -> r.name(name));
        Assertions.assertEquals("updated1", resp.parameter().value(), "PutParameterOverwrite: value not updated");
    }

    private void getParameterHistory(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmParam1");
        var resp = ssm().getParameterHistory(r -> r.name(name));
        Assertions.assertNotEmpty(resp.parameters(), "GetParameterHistory: no history entries");
    }

    private void putMultipleParameters(TestContext ctx) throws Exception {
        // Verify both params from setup are accessible.
        String n1 = ctx.getString("ssmParam1");
        String n2 = ctx.getString("ssmParam2");
        var resp = ssm().getParameters(r -> r.names(n1, n2));
        Assertions.assertEquals(2, resp.parameters().size(), "PutMultipleParameters: expected 2 params");
    }

    private void getParameters(TestContext ctx) throws Exception {
        String n1 = ctx.getString("ssmParam1");
        String n2 = ctx.getString("ssmParam2");
        var resp = ssm().getParameters(r -> r.names(n1, n2));
        Assertions.assertEquals(2, resp.parameters().size(), "GetParameters: expected 2 parameters");
    }

    private void describeParameters(TestContext ctx) throws Exception {
        var resp = ssm().describeParameters(r -> r.maxResults(10));
        Assertions.assertNotNull(resp.parameters(), "DescribeParameters: parameters is null");
    }

    private void tagParameter(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmParam1");
        ssm().addTagsToResource(r -> r
                .resourceType(ResourceTypeForTagging.PARAMETER)
                .resourceId(name)
                .tags(software.amazon.awssdk.services.ssm.model.Tag.builder()
                        .key("env").value("test").build()));
    }

    private void listSsmTagsForResource(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmParam1");
        var resp = ssm().listTagsForResource(r -> r
                .resourceType(ResourceTypeForTagging.PARAMETER)
                .resourceId(name));
        Assertions.assertNotNull(resp.tagList(), "ListSSMTagsForResource: tagList is null");
    }

    private void deleteParameters(TestContext ctx) throws Exception {
        String n1 = ctx.getString("ssmParam1");
        String n2 = ctx.getString("ssmParam2");
        ssm().deleteParameters(r -> r.names(n1, n2));
        // Clear ctx so teardown does not attempt double-delete.
        ctx.set("ssmParam1", null);
        ctx.set("ssmParam2", null);
    }

    // ── ssm-secure ────────────────────────────────────────────────────────────

    private void setupSecure(TestContext ctx) {
        ctx.set("ssmSecureParam", "/" + ctx.runId() + "/secure");
    }

    private void putSecureStringParameter(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmSecureParam");
        ssm().putParameter(r -> r.name(name).value("s3cr3t").type(ParameterType.SECURE_STRING));
    }

    private void getSecureStringParameter(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmSecureParam");
        var resp = ssm().getParameter(r -> r.name(name).withDecryption(true));
        Assertions.assertEquals("s3cr3t", resp.parameter().value(), "GetSecureStringParameter: value mismatch");
    }

    private void getSecureStringWithoutDecryption(TestContext ctx) throws Exception {
        String name = ctx.getString("ssmSecureParam");
        var resp = ssm().getParameter(r -> r.name(name).withDecryption(false));
        Assertions.assertNotNull(resp.parameter().value(), "GetSecureStringWithoutDecryption: value is null");
        // The value should be the ciphertext placeholder, not the plaintext.
        Assertions.assertNotEquals("s3cr3t", resp.parameter().value(),
                "GetSecureStringWithoutDecryption: expected encrypted/redacted value");
    }

    // ── ssm-path ──────────────────────────────────────────────────────────────

    private void setupPath(TestContext ctx) throws Exception {
        String prefix = "/" + ctx.runId() + "/app";
        ssm().putParameter(r -> r.name(prefix + "/db-url").value("jdbc://localhost").type(ParameterType.STRING));
        ssm().putParameter(r -> r.name(prefix + "/db-pass").value("secret").type(ParameterType.STRING));
        ssm().putParameter(r -> r.name(prefix + "/nested/key").value("nested-value").type(ParameterType.STRING));
        ctx.set("ssmPathPrefix", prefix);
    }

    private void teardownPath(TestContext ctx) {
        String prefix = ctx.getString("ssmPathPrefix");
        if (prefix == null) return;
        deleteParamSilently(prefix + "/db-url");
        deleteParamSilently(prefix + "/db-pass");
        deleteParamSilently(prefix + "/nested/key");
    }

    private void getParametersByPath(TestContext ctx) throws Exception {
        String prefix = ctx.getString("ssmPathPrefix");
        var resp = ssm().getParametersByPath(r -> r.path(prefix + "/").recursive(false));
        Assertions.assertGreaterThanOrEqual(2, resp.parameters().size(),
                "GetParametersByPath: expected >= 2 parameters at " + prefix + "/");
    }

    private void getParametersByPathRecursive(TestContext ctx) throws Exception {
        String prefix = ctx.getString("ssmPathPrefix");
        var resp = ssm().getParametersByPath(r -> r.path(prefix + "/").recursive(true));
        Assertions.assertGreaterThanOrEqual(3, resp.parameters().size(),
                "GetParametersByPathRecursive: expected >= 3 parameters (incl. nested)");
    }

    private void getParametersByPathPaginated(TestContext ctx) throws Exception {
        String prefix = ctx.getString("ssmPathPrefix");
        var page1 = ssm().getParametersByPath(r -> r.path(prefix + "/").recursive(true).maxResults(1));
        Assertions.assertGreaterThanOrEqual(1, page1.parameters().size(),
                "GetParametersByPathPaginated: page 1 should have at least 1 parameter");
        if (page1.nextToken() != null) {
            var page2 = ssm().getParametersByPath(r -> r.path(prefix + "/")
                    .recursive(true).maxResults(1).nextToken(page1.nextToken()));
            Assertions.assertGreaterThanOrEqual(1, page2.parameters().size(),
                    "GetParametersByPathPaginated: page 2 should have at least 1 parameter");
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void deleteParamSilently(String name) {
        if (name == null) return;
        try { ssm().deleteParameter(r -> r.name(name)); } catch (Exception ignored) {}
    }
}
